Add NewtonianEuler-specialized minmod limiter

fmahebert opened the pull request:
## Proposed changes

Adds a simple limiter specialized to the NewtonianEuler system. This means the limiter
- can act on the characteristic variables, as recommended
- can follow the limiting with an additional flattening step that helps avoid negative densities and pressures


### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Add Minmod limiter specialized to NewtonianEuler

fmahebert committed:
Add VariablesToLimit enum for NewtonianEuler limiting

fmahebert committed:
Add helper to apply a limiter to NewtonianEuler char vars

fmahebert committed:
Factor code in NewtonianEuler limiter test

This code will be reused by an upcoming test.

fmahebert committed:
Rename unit_normal to unit_vector in limiter helper

In the helper code for limiting the NewtonianEuler characteristic
variables, the unit vector used to compute the characteristic
transformation is not a normal to anything... it is just a unit
vector.

fmahebert committed:
Switch to system-specific Minmod in NewtonianEuler exec

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/VariablesToLimit.hpp:
```diff
@@ -0,0 +1,37 @@
+
+#include <ostream>
+
+namespace Options {
```
`/// cond` for forward decls

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.hpp:
```diff
@@ -0,0 +1,188 @@
+///
+/// Implements the three minmod-based generalized slope limiters from
+/// \cite Cockburn1999 Sec. 2.4: \f$\Lambda\Pi^1\f$, \f$\Lambda\Pi^N\f$, and
+/// MUSCL. See the documentation of the system-agnostic Minmod limiter for a
```
Any chance you could link to the namespace so it's easier to find?

fmahebert replied:
I tried putting in `Limiters::Minmod` (links to NewtonianEuler version still) or `::Limiters::Minmod` (links to nothing) or trying different link/ref commands, but couldn't get it to link to the generic Minmod. Any suggestions of other things to try?

moxcodes replied:
oh, I see... the actual `Limiters::Minmod` documentation is dox on a template specialization and doxygen is bad at template specializations.

My one suggestion, that you can feel free to ignore, is to move the documentation of ::Limiters::Minmod in `DiscontinuousGalerkin/Minmod.hpp` from the template specialization to the declaration on line 48.

fmahebert replied:
This is the move! I'm adding a commit

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.cpp:
```diff
@@ -0,0 +1,282 @@
+      typename CharacteristicVarsMinmod::PackagedData,
+      boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>
+      neighbor_char_data{};
+  for (const auto& kv : neighbor_data) {
```
could use structured binding here.

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.cpp:
```diff
@@ -0,0 +1,282 @@
+          const gsl::not_null<Scalar<DataVector>*> char_v_plus,
+          const Matrix& left_eigenvectors) noexcept -> bool {
+    // Convert neighbor data to characteristics
+    for (const auto& kv : neighbor_data) {
```
structured binding

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.cpp:
```diff
@@ -0,0 +1,282 @@
+    const Scalar<DataVector>& energy_density, const Mesh<VolumeDim>& mesh,
+    const std::array<double, VolumeDim>& element_size,
+    const OrientationMap<VolumeDim>& orientation_map) const noexcept {
+  const ConservativeVarsMinmod minmod(minmod_type_, tvb_constant_,
```
Does it make sense for the `minmod` object to be a member variable, given that it's always constructed from the saved options?

nilsdeppe commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.cpp:
```diff
@@ -0,0 +1,283 @@
+        get(*energy_density) / get(*mass_density_cons) -
+        0.5 * get(dot_product(*momentum_density, *momentum_density)) /
+            square(get(*mass_density_cons))};
+    const auto pressure = equation_of_state.pressure_from_density_and_energy(
```
can you do the pressure computation inside the `ASSERT`? Maybe wrap this entire if block in an `ifdef SPECTRE_DEBUG`?

fmahebert replied:
Good call, I'll do something like this

nilsdeppe reviewed (dismissed):
Overall LGTM, one suggestion that you can squash in immediately if you'd like to include it :)

moxcodes commented on src/Evolution/Systems/NewtonianEuler/Limiters/Minmod.cpp:
```diff
@@ -0,0 +1,283 @@
+        << vars_to_limit_);
+  }
+
+  if (apply_flattener_) {
```
This appears to apply the flattener regardless of whether the limiter activated. I'm not familiar with the flattener technique, but the documentation refered to it an operation on the 'post-limiter solution', which could be interpreted to indicate that it would only be used to modify cases that were limited.
It might be helpful to have the documentation give more details about which cases apply the flattener.

moxcodes commented on tests/Unit/Evolution/Systems/NewtonianEuler/Limiters/Test_Minmod.cpp:
```diff
@@ -0,0 +1,629 @@
+// characteristic variables is tedious to do thoroughly, because we need to
+// call the generic limiter multiple times using characteristic fields w.r.t.
+// multiple different unit vectors. In essence, the test would need to fully
+// reimplement the specialized limiter. To avoid the test become so tedious, we
```
[nitpick] `test become` -> `test becoming` or similar grammatical fix

moxcodes reviewed (commented):
Only small requests from me, feel free to squash any changes

fmahebert commented:
You invited me to squash, but I posted the fixups because
- I did a bit of rephrasings beyond what was asked
- I fixed some bugs in the test. The test checks that the specialized and generic minmod match, which they did and still do. But the data passed to the limiter for this check wasn't as I intended it to be, and now it is.

moxcodes reviewed (commented):
Looks good -- you have my go-ahead to squash. If you choose to do try the dox tweak to `DiscontinuousGalerkin/Minmod.hpp`, I think that ought to just be its own commit.

nilsdeppe reviewed (commented):
I'm also happy for you to squash

fmahebert committed:
Fix Doxygen documentation for Minmod limiter

By moving the Doxygen from the template specialization to the
declaration, we help Doxygen understand how to link to this
documentation from elsewhere in the code.

fmahebert commented:
Squashed and rebased. I added a commit per @moxcodes's suggestion that moves the documentation of the generic Minmod limiter so that Doxygen can link to it
